Rework odoc rules

lubegasimon opened the pull request:
This PR adds `odoc link` as part of [the documentation generation phases](https://ocaml.github.io/odoc/driver.html#document-generation-phases) and stops `dune` from generating documentation for hidden modules (those with double underscores in their names).

lubegasimon commented:
Ping: Anything blocking this PR?

jonludlam commented:
I think you need to satisfy at least the DCO bot, and there's a conflict. Additionally, I had a go at removing `odocl_alias` and it all seems to work fine: https://github.com/jonludlam/dune/commit/257f58d178caa775179e23700d9362daec31e9ec - I think it's not strictly necessary so unless there's a compelling reason why we need the alias it seems better to remove it. Having said that I'm not sure on whether dune ordinarily has helpful aliases around - perhaps @rgrinberg might have an opinion?

lubegasimon commented:
> jonludlam@257f58d
I will make fixes and squash commits, thanks!

lubegasimon commented:
@rgrinberg any blockers to get this merged (except the conflicts)?

rgrinberg commented:
There are no blockers yet but I haven't reviewed your latest changes. I'll do so today.

rgrinberg commented:
This PR is still special casing modules with `__` in their name. This behavior is not desirable for the reasons I've outlined above.

lubegasimon commented:
> This PR is still special casing modules with `__` in their name. This behavior is not desirable for the reasons I've outlined above.

I interpret this as asking authors it concerns to rewrite their dune files, for which I’m not sure how many are willing to do so urgently in favour of this PR.

/c.c @aantron

jonludlam commented:
I believe right now there won't be any problems, because odoc has some compatibility code in to write an empty file when asked to generate an HTML file for a hidden module. When we remove this I suspect there will be problems because dune will be expecting an output file that won't get written.

lubegasimon committed:
introduce and add odoc-link rules

Signed-off-by: lubegasimon <lubegasimon73@gmail.com>

Remove odocl_alias

Signed-off-by: Jon Ludlam <jon@recoil.org>

lubegasimon committed:
stop dune from generating docs for hidden modules

Signed-off-by: lubegasimon <lubegasimon73@gmail.com>

lubegasimon committed:
test if dune does not generate docs for hidden modules

Signed-off-by: lubegasimon <lubegasimon73@gmail.com>

fix CI failures by sorting test results

Signed-off-by: lubegasimon <lubegasimon73@gmail.com>

test if odocl files are formed

Signed-off-by: lubegasimon <lubegasimon73@gmail.com>

more CI failure fixes by sorting test results

Signed-off-by: lubegasimon <lubegasimon73@gmail.com>

remove redundant compiler invocation

Signed-off-by: lubegasimon <lubegasimon73@gmail.com>

lubegasimon commented:
PR ready for another review!

jonludlam commented:
Looking good! I still think we should remove the odocl_alias as I suggested above though.

lubegasimon commented:
> Looking good! I still think we should remove the odocl_alias as I suggested above though.



> Looking good! I still think we should remove the odocl_alias as I suggested above though.

Done! I think @jonludlam will need to sign off the commit I cherry-picked to pass the failing check.

lubegasimon commented:
It seems the only blocker now is the failing macos-workflow
/c.c @rgrinberg

jonludlam commented:
Agreed, aside from that this is good to merge as far as I can tell.

rgrinberg commented on test/blackbox-tests/test-cases/odoc/odoc-wrapped-lib.t/run.t:
```diff
@@ -0,0 +1,23 @@
+This test generates documentation for non-hidden modules only for a library:
+
+  $ ocamlc -c -bin-annot foo.ml
```
Why do we need this `ocamlc` call?

lubegasimon replied:
Surely, that’s a redundant call! Good catch!
